File: Orix-App/backend/app/services/support_service.py
import uuid
import logging
from typing import List, Optional, Dict, Any

from app.database import get_db
from app.models.support import SupportTicket, TicketMessage, CreateTicketRequest
from app.services.notification_service import NotificationService
from app.services.audit_service import AuditService
from app.models.notification import NotificationType
from app.utils.timezone_utils import utc_now, format_ist

logger = logging.getLogger(__name__)

class SupportService:
    """Service for managing support tickets and reports."""
    
    def __init__(self):
        self.notification_service = NotificationService()
        self.audit_service = AuditService()
    
    async def create_ticket(self, user_id: str, request: CreateTicketRequest) -> SupportTicket:
        """Create a new support ticket and notify admins via Telegram."""
        db = get_db()
        
        ticket_id = f"TICKET-{str(uuid.uuid4())[:8].upper()}"
        
        # Initial message from user description
        initial_message = TicketMessage(
            sender_id=user_id,
            content=request.description,
            timestamp=utc_now(),
            is_admin=False
        )
        
        ticket = SupportTicket(
            ticket_id=ticket_id,
            user_id=user_id,
            type=request.type,
            subject=request.subject,
            description=request.description,
            status="open",
            priority=request.priority,
            reported_user_id=request.reported_user_id,
            messages=[initial_message],
            created_at=utc_now(),
            updated_at=utc_now()
        )
        
        await db.support_tickets.insert_one(ticket.model_dump())
        
        # Log to Telegram Admin Channel
        await self._notify_admins_telegram(ticket)
        
        return ticket

    async def get_user_tickets(self, user_id: str) -> List[SupportTicket]:
        """Get all tickets for a user."""
        db = get_db()
        cursor = db.support_tickets.find({"user_id": user_id}).sort("updated_at", -1)
        return [SupportTicket(**doc) async for doc in cursor]

    async def get_ticket(self, ticket_id: str) -> Optional[SupportTicket]:
        """Get a single ticket by ID."""
        db = get_db()
        doc = await db.support_tickets.find_one({"ticket_id": ticket_id})
        return SupportTicket(**doc) if doc else None

    async def get_all_open_tickets(self) -> List[SupportTicket]:
        """Get all open tickets for admins."""
        db = get_db()
        cursor = db.support_tickets.find(
            {"status": {"$in": ["open", "in_progress"]}}
        ).sort("created_at", -1)
        
        tickets = []
        async for doc in cursor:
            try:
                tickets.append(SupportTicket(**doc))
            except Exception as e:
                logger.warning(f"Failed to parse ticket {doc.get('ticket_id', 'unknown')}: {e}")
                continue
        return tickets

    async def add_user_message(self, ticket_id: str, user_id: str, content: str) -> Optional[SupportTicket]:
        """Add a message from the user to the ticket."""
        db = get_db()
        
        ticket = await self.get_ticket(ticket_id)
        if not ticket or ticket.user_id != user_id:
            return None
            
        message = TicketMessage(
            sender_id=user_id,
            content=content,
            timestamp=utc_now(),
            is_admin=False
        )
        
        update_result = await db.support_tickets.find_one_and_update(
            {"ticket_id": ticket_id},
            {
                "$push": {"messages": message.model_dump()},
                "$set": {
                    "updated_at": utc_now(),
                    "status": "open" if ticket.status == "resolved" else ticket.status 
                }
            },
            return_document=True
        )
        
        if update_result:
            updated_ticket = SupportTicket(**update_result)
            # Notify admins of new reply
            await self._notify_admins_reply(updated_ticket, message)
            # Auto-reply acknowledgement to user
            await self._send_auto_reply(updated_ticket)
            return updated_ticket
            
        return None
    
    async def _send_auto_reply(self, ticket: SupportTicket):
        """Send auto-reply acknowledgement to user when they send a message."""
        try:
            await self.notification_service.send_notification(
                user_id=ticket.user_id,
                notification_type=NotificationType.SYSTEM,
                title="Message Received",
                body=f"We received your message on ticket {ticket.ticket_id}. An admin will respond shortly.",
                data={"type": "ticket_auto_reply", "ticket_id": ticket.ticket_id},
                send_push=True
            )
        except Exception as e:
            logger.warning(f"Failed to send auto-reply for ticket {ticket.ticket_id}: {e}")

    async def add_admin_reply(self, ticket_id: str, admin_id: str, content: str) -> Optional[SupportTicket]:
        """Add a reply from an admin via Telegram."""
        db = get_db()
        
        ticket = await self.get_ticket(ticket_id)
        if not ticket:
            return None
            
        message = TicketMessage(
            sender_id="admin",
            sender_name=f"Admin ({admin_id})",
            content=content,
            timestamp=utc_now(),
            is_admin=True
        )
        
        update_result = await db.support_tickets.find_one_and_update(
            {"ticket_id": ticket_id},
            {
                "$push": {"messages": message.model_dump()},
                "$set": {
                    "updated_at": utc_now(),
                    "status": "in_progress" # Mark in progress when admin replies
                }
            },
            return_document=True
        )
        
        if update_result:
            updated_ticket = SupportTicket(**update_result)
            
            # Notify user
            await self.notification_service.send_notification(
                user_id=ticket.user_id,
                notification_type=NotificationType.SYSTEM,
                title="Support Update",
                body=f"New reply on ticket {ticket.ticket_id}: {content[:50]}...",
                data={"type": "ticket_update", "ticket_id": ticket_id},
                send_push=True
            )
            
            return updated_ticket
            
        return None

    async def _notify_admins_telegram(self, ticket: SupportTicket):
        """Send formatted notification to Telegram log channel."""
        try:
            timestamp = format_ist(utc_now())
            
            type_tag = "[ISSUE]" if ticket.type == "issue" else "[FEATURE]" if ticket.type == "feature" else "[ALERT]"
            
            msg = (
                f"{type_tag} *NEW SUPPORT TICKET*\n"
                f"`{timestamp}`\n"
                f"ID: `{ticket.ticket_id}`\n"
                f"User: `{ticket.user_id}`\n"
                f"Type: *{ticket.type.upper()}*\n"
                f"Subject: {ticket.subject}\n"
                f"Desc: {ticket.description}\n\n"
                f"Reply usage: `/reply {ticket.ticket_id} <message>`"
            )
            
            await self.audit_service._send_telegram_log(msg)
        except Exception:
            pass

    async def _notify_admins_reply(self, ticket: SupportTicket, message: TicketMessage):
        """Notify admins that user replied."""
        try:
            timestamp = format_ist(utc_now())
            
            msg = (
                f"[USER REPLY]\n"
                f"`{timestamp}` | Ticket `{ticket.ticket_id}`\n"
                f"User: `{ticket.user_id}`\n"
                f"Message: {message.content}\n\n"
                f"Reply usage: `/reply {ticket.ticket_id} <message>`"
            )
            
            await self.audit_service._send_telegram_log(msg)
        except Exception:
            pass

    async def close_ticket(self, ticket_id: str, closed_by: str, reason: str = None) -> Optional[SupportTicket]:
        """Close a support ticket."""
        db = get_db()
        
        ticket = await self.get_ticket(ticket_id)
        if not ticket:
            return None
            
        update_result = await db.support_tickets.find_one_and_update(
            {"ticket_id": ticket_id},
            {
                "$set": {
                    "status": "closed",
                    "updated_at": utc_now(),
                    "resolved_at": utc_now(),
                    "resolved_by": closed_by
                }
            },
            return_document=True
        )
        
        if update_result:
            updated_ticket = SupportTicket(**update_result)
            
            # Notify User if closed by admin
            if closed_by != ticket.user_id:
                await self.notification_service.send_notification(
                    user_id=ticket.user_id,
                    notification_type=NotificationType.SYSTEM,
                    title="Ticket Closed",
                    body=f"Your ticket {ticket_id} has been closed by support.",
                    data={"type": "ticket_closed", "ticket_id": ticket_id},
                    send_push=True
                )
            
            timestamp = format_ist(utc_now())
            msg = (
                f"[TICKET CLOSED]\n"
                f"`{timestamp}` | Ticket `{ticket_id}`\n"
                f"Closed by: `{closed_by}`\n"
                f"Reason: {reason or 'Resolved'}"
            )
            await self.audit_service._send_telegram_log(msg)
            
            return updated_ticket
        
        return None

File: Orix-App/frontend/lib/features/profile/help_support/support_service.dart
import '../../../core/services/api_service.dart';
import '../../../core/config/api_config.dart';
import '../../../core/utils/time_utils.dart';

class SupportTicket {
  final String ticketId;
  final String userId;
  final String type;
  final String subject;
  final String description;
  final String status;
  final List<TicketMessage> messages;
  final DateTime createdAt;
  final DateTime updatedAt;

  SupportTicket({
    required this.ticketId,
    required this.userId,
    required this.type,
    required this.subject,
    required this.description,
    required this.status,
    required this.messages,
    required this.createdAt,
    required this.updatedAt,
  });

  factory SupportTicket.fromJson(Map<String, dynamic> json) {
    return SupportTicket(
      ticketId: json['ticket_id'],
      userId: json['user_id'],
      type: json['type'],
      subject: json['subject'],
      description: json['description'],
      status: json['status'],
      messages: (json['messages'] as List)
          .map((m) => TicketMessage.fromJson(m))
          .toList(),
      createdAt: TimeUtils.parseUtc(json['created_at']),
      updatedAt: TimeUtils.parseUtc(json['updated_at']),
    );
  }
}

class TicketMessage {
  final String senderId;
  final String? senderName;
  final String content;
  final DateTime timestamp;
  final bool isAdmin;

  TicketMessage({
    required this.senderId,
    this.senderName,
    required this.content,
    required this.timestamp,
    required this.isAdmin,
  });

  factory TicketMessage.fromJson(Map<String, dynamic> json) {
    return TicketMessage(
      senderId: json['sender_id'],
      senderName: json['sender_name'],
      content: json['content'],
      timestamp: TimeUtils.parseUtc(json['timestamp']),
      isAdmin: json['is_admin'] ?? false,
    );
  }
}

class SupportService {
  final ApiService _apiService;

  SupportService(this._apiService);

  Future<SupportTicket> createTicket({
    required String type,
    required String subject,
    required String description,
    String? reportedUserId,
  }) async {
    final response = await _apiService.post(
      '/support/tickets',
      body: {
        'type': type,
        'subject': subject,
        'description': description,
        'reported_user_id': reportedUserId,
      },
    );

    if (!response.success || response.data == null) {
      throw Exception(response.error ?? 'Failed to create ticket');
    }

    return SupportTicket.fromJson(response.data);
  }

  Future<List<SupportTicket>> getUserTickets() async {
    final response = await _apiService.get('/support/tickets');

    if (!response.success || response.data == null) {
      throw Exception(response.error ?? 'Failed to fetch tickets');
    }

    return (response.data as List)
        .map((t) => SupportTicket.fromJson(t))
        .toList();
  }

  Future<SupportTicket> getTicket(String ticketId) async {
    final response = await _apiService.get('/support/tickets/$ticketId');

    if (!response.success || response.data == null) {
      throw Exception(response.error ?? 'Failed to fetch ticket details');
    }

    return SupportTicket.fromJson(response.data);
  }

  Future<SupportTicket> sendMessage(String ticketId, String content) async {
    final response = await _apiService.post(
      '/support/tickets/$ticketId/message',
      body: {'content': content},
    );

    if (!response.success || response.data == null) {
      throw Exception(response.error ?? 'Failed to send message');
    }

    return SupportTicket.fromJson(response.data);
  }

  Future<SupportTicket> closeTicket(String ticketId) async {
    final response = await _apiService.post(
      '/support/tickets/$ticketId/close',
      body: {},
    );

    if (!response.success || response.data == null) {
      throw Exception(response.error ?? 'Failed to close ticket');
    }

    return SupportTicket.fromJson(response.data);
  }

  Future<List<UserSearchResult>> searchUsers(String query) async {
    final response = await _apiService.get('/users/search?query=$query');

    if (!response.success || response.data == null) {
      return [];
    }

    return (response.data as List)
        .map((u) => UserSearchResult.fromJson(u))
        .toList();
  }
}

class UserSearchResult {
  final String userId;
  final String displayName;
  final String email;
  final String? photoUrl;

  UserSearchResult({
    required this.userId,
    required this.displayName,
    required this.email,
    this.photoUrl,
  });

  factory UserSearchResult.fromJson(Map<String, dynamic> json) {
    String? photoUrl = json['photo_url'] ?? json['photoUrl'];
    if (photoUrl != null && photoUrl.startsWith('/')) {
      photoUrl = '${ApiConfig.baseUrl}$photoUrl';
    }

    return UserSearchResult(
      userId: json['user_id'] ?? json['userId'] ?? '',
      displayName: json['display_name'] ?? json['displayName'] ?? 'Unknown',
      email: json['email'] ?? 'No Email',
      photoUrl: photoUrl,
    );
  }
}
